### ISSUE TITLE
fix(phi3): select LongRoPE short or long factors by sequence position and read mscale overrides

### ISSUE THREAD

**Issue by inureyes** [lablup/mlxcel#1358]
## Summary

Phi-3.5 / Phi-4-mini / Phi-4-multimodal ship `rope_scaling.type = "longrope"` (or `"su"`) with two per-dimension factor lists. The trained model uses `short_factor` while the sequence fits in `original_max_position_embeddings` (4096) and switches to `long_factor` beyond it, with an attention-magnitude scale that may be overridden per list (`short_mscale` / `long_mscale`). mlxcel builds a single frequency table from `long_factor` and uses it at every position. On Phi-3.5-mini `long_factor` reaches 64.8 on the low-frequency pairs while `short_factor` stays within 1.0..1.03, so every prompt under 4096 tokens is rotated with the wrong table. This change keeps both tables, picks per forward pass from `offset + seq_len`, and honors the optional mscale keys.

## Current behavior

`src/models/phi3.rs:325-365` `configure_su_rope` reads `long_factor` only:

```rust
let long_factor = match &scaling.long_factor { Some(f) => f, None => return };
...
freqs[i] = (long_factor[i] * base.powf(exponent)) as f32;
self.su_rope_freqs = Some(...);
```

`short_factor` is parsed (`src/models/phi3.rs:81-82`) and never read. The single table feeds both the fused quantized path `forward_fused_qkv_split_su_scaled_rope` (`src/models/phi3.rs:166-176`) and the graph fallback `prepare_qkv_with_rope` (`src/models/phi3.rs:238-248`). The scale is always the default `sqrt(1 + ln(max_pos / orig) / ln(orig))` (`src/models/phi3.rs:352-360`); `short_mscale` / `long_mscale` are not parsed. `original_max_position_embeddings` is read only from inside the `rope_scaling` block (`src/models/phi3.rs:352`), while the checkpoints put it at the top level (`models/phi-3.5-mini-4bit/config.json`: top-level `"original_max_position_embeddings": 4096`), so the value is right only because the default is 4096.

Phi-4-multimodal (`src/models/phi4mm.rs:20-21` reuses `phi3::ModelArgs` / `Phi3Model`) inherits the same behavior. Reference configs: `microsoft/Phi-3.5-mini-instruct` `short_factor[0..4] = [1.0, 1.02, 1.03, 1.03]`, `long_factor[0..4] = [1.08, 1.11, 1.14, 1.34]`, `long_factor[45..48] = [64.51, 64.53, 64.84]`; `microsoft/Phi-4-multimodal-instruct` `short_factor = [1.0] * 48`, `long_factor` up to 47.77, `partial_rotary_factor 0.75`.

## Expected behavior

With `d = rope_dims`, `base = rope_theta`, `L = original_max_position_embeddings`, `M = max_position_embeddings`:

```
freq_i          = base ^ (2i / d),            i in 0 .. d/2
short_freqs_i   = short_factor[i] * freq_i
long_freqs_i    = long_factor[i]  * freq_i
factor          = M / L
default_scale   = 1.0 if factor <= 1 else sqrt(1 + ln(factor) / ln(L))
short_scale     = short_mscale if present else default_scale
long_scale      = long_mscale  if present else default_scale

per forward pass with cache offset `o` and `s` new tokens:
    position_end = max(total prompt length, o + s)
    use_long     = position_end > L
    freqs = long_freqs if use_long else short_freqs
    scale = long_scale if use_long else short_scale
    q[..., :d] *= scale;  k[..., :d] *= scale
    q = rope(q, d, freqs = freqs, offset = o);  k likewise
```

The switch is per sequence, not per pass and not per token. **Corrected during implementation:** the original wording here selected from `o + s` alone, which is right for HuggingFace (it always prefills a whole prompt in one pass) and wrong for mlxcel, which prefills a long prompt in chunks (`DEFAULT_PREFILL_CHUNK = 2048` on the CLI, `--prefill-chunk-size` on the server). Under the per-pass rule, chunks 1 and 2 of a 5136-token prompt end at 2048 and 4096 and take `short_freqs` while chunk 3 takes `long_freqs`, so one KV cache holds keys built from two tables and greedy output degenerates into repetition within a few tokens (measured against a patched copy of mlx-lm's `SuScaledRoPE`). So `position_end` is the total prompt length known when the prefill starts, and every chunk of one prompt resolves the same table from it; a single-pass prefill needs nothing extra because its own `o + s` is already the sequence length. Decode still flips to `long_freqs` once `o + 1 > L`, and batched decode rows use their own `offset` per row. `original_max_position_embeddings` is read from the `rope_scaling` block first, then the top level, default 4096.

Note also that when decode crosses `L`, transformers drops the cache and re-encodes the whole sequence with the long table (`prepare_inputs_for_generation`, the branch on `input_ids.shape[1] >= original_max_position_embeddings + 1` and `past_length <= original_max_position_embeddings`). The flip-in-place above keeps short-table keys in the cache and is an approximation of that, deliberately, so transformers parity is not claimed for a generation that crosses `L` mid-decode.

## Implementation plan

1. `src/models/phi3.rs` `RopeScaling` (`src/models/phi3.rs:66-83`): add `#[serde(default)] pub short_mscale: Option<f32>`, `pub long_mscale: Option<f32>`. `ModelArgs`: add top-level `#[serde(default)] pub original_max_position_embeddings: Option<usize>` and a `fn original_max_position_embeddings(&self) -> usize` resolving block, then top level, then 4096.
2. `Attention` fields (`src/models/phi3.rs:140-147`): replace `su_rope_freqs` / `su_rope_scale` / `su_rope_scale_arr` with `su_short: Option<SuRopeTable>`, `su_long: Option<SuRopeTable>` where `struct SuRopeTable { freqs: UniquePtr<MlxArray>, scale: f32, scale_arr: Option<UniquePtr<MlxArray>> }`, plus `su_original_max: i32`. `configure_su_rope` builds both tables (require both lists of length `d/2`; a config with only `long_factor` uses it for both, matching today).
3. Add `fn su_table_for(&self, offset: i32, seq_len: i32) -> Option<&SuRopeTable>` implementing the rule above; call it at the top of `Attention::forward` (`src/models/phi3.rs:160-180`) and pass the selected table's `freqs` / `scale` / `scale_arr` into `forward_fused_qkv_split_su_scaled_rope` and `prepare_qkv_with_rope` / `scale_su_rope_rotary_prefix` (`src/models/phi3.rs:253-275`). The batched decode path (if any `fast_rope_batched` call exists in this file after the Phi paged work) selects per row.
4. Phi-3 Vision (`src/models/phi3` consumers in `src/vision/phi3_vl.rs`) and Phi4MM (`src/models/phi4mm.rs`) need no change beyond recompiling; `type: "su"` and `type: "longrope"` both take this path as today.
5. `docs/supported-models.md` Phi-3 / Phi4MM entries: state the position-dependent table selection and that the oracle for token-exact comparison below 4096 tokens must be an implementation that selects `short_factor` there (the transformers modeling code does); an oracle that applies `long_factor` unconditionally is not a valid reference for short prompts.

## Validation

(a) Unit tests in `src/models/phi3_tests.rs`:

- `su_rope_builds_short_and_long_tables`: synthetic `d = 8`, `short = [1,1,1,1]`, `long = [2,4,8,16]`; assert `short_freqs == freq`, `long_freqs == long * freq`.
- `su_table_switches_at_original_max`: `L = 16`: `(offset 0, len 16)` selects short, `(offset 0, len 17)` selects long, `(offset 15, len 1)` short, `(offset 16, len 1)` long.
- `mscale_override_wins`: `short_mscale = 1.5`, `long_mscale = None`, `M = 8 * L` gives short scale `1.5` and long scale `sqrt(1 + ln(8) / ln(L))`.
- `fused_and_graph_su_paths_agree_per_table`: quantized layer, both tables, RMS < 1e-3 in f16 between the fused and graph outputs.

(b) Real checkpoints: `mlx-community/Phi-3.5-mini-instruct-4bit` (local `models/phi-3.5-mini-4bit`) and `microsoft/Phi-4-multimodal-instruct` (the documented Phi4MM checkpoint).

```
./target/release/mlxcel generate -m models/phi-3.5-mini-4bit -p "<a 1024-token prompt ending with a question about its first sentence>" -n 64
./target/release/mlxcel generate -m models/phi-3.5-mini-4bit -p "<a 5000-token prompt ending with a question about its first sentence>" -n 64
```

Acceptance: on the 1024-token prompt greedy output is token-exact against the transformers model (which uses `short_factor` there); on the 5000-token prompt output is token-exact against the same oracle and unchanged from the current mlxcel output (both use `long_factor` there). Phi4MM: one image prompt under 4096 tokens, token-exact against the transformers model.

## Acceptance criteria

- [x] Short and long LongRoPE tables are built, and selected from the total prompt length rather than one pass's `offset + seq_len` (see the correction above)
- [x] `short_mscale` / `long_mscale` override the default scale when present; top-level `original_max_position_embeddings` is honored
- [x] Fused and graph paths agree for both tables
- [ ] Phi-3.5-mini token-exact vs the transformers oracle under 4096 tokens and unchanged above it (real-checkpoint gate, run outside the implementation unit)
- [x] docs/supported-models.md updated
- [ ] cargo test --workspace --profile test-fast --features metal,accelerate passes (the narrow `--lib models::phi3::tests`, `-p mlxcel-core --lib prefill_span` and `--lib models::rope_utils` runs all pass; the full workspace run is outside the implementation unit)
- [ ] cargo clippy --workspace --all-targets -- -D warnings and cargo fmt --all -- --check pass (`cargo fmt --all -- --check` and `cargo clippy --lib --tests -- -D warnings` on both changed crates pass; `--all-targets` is outside the implementation unit)

## Out of scope

- Phi-3-small (`phi3small`) and Phi-3.5-MoE, which have their own attention modules; audit them separately if they carry `longrope`.

**Comment by inureyes** [comment]
Implemented in #1580, with one correction to the specification in the body.

The rule "select per forward pass from `offset + seq_len > original_max_position_embeddings`" is what HuggingFace transformers computes, and it is correct there because transformers prefills a whole prompt in one pass. mlxcel does not: `src/lib/mlxcel-core/src/generate.rs` sets `DEFAULT_PREFILL_CHUNK = 2048` for the single-sequence CLI and bench path, and the server's batch scheduler splits a prefill across ticks with its own `--prefill-chunk-size`. Under the per-pass rule a 5136-token prompt at the default chunk size takes the short table for chunks 1 and 2 (they end at 2048 and 4096, both inside the trained context) and the long table for chunk 3, so one KV cache ends up holding keys rotated with two different tables and the greedy output degenerates into repetition within a few tokens.

The selection is therefore made from the total prompt length known when the prefill starts: every chunk of a prompt longer than `L` uses the long table, every chunk of a shorter prompt uses the short table. A new `mlxcel_core::prefill_span` carries that length from the driver that splits the prefill to the model that reads it, as a thread-local announcement scoped by an RAII guard to one chunk's forward call, so the decode batches the server scheduler interleaves between two chunks cannot read it. Single-pass prefills announce nothing, because their own `offset + seq_len` is already the sequence length, and the reader takes the larger of the two so an under-announcing driver cannot make a long sequence look short. The issue body and its pseudocode have been updated to match.

Two other things worth recording here. First, mlx-lm is not a valid oracle for this change: its `SuScaledRotaryEmbedding` accepts `short_factor` and `short_mscale` as constructor arguments and then builds its frequencies from `long_factor` and its scale from `long_mscale` alone, so it applies the long table at every position, which is exactly what mlxcel did before this fix. The reference that matches the trained model is `Phi3LongRoPEScaledRotaryEmbedding` in transformers, and that is stated in the code, the PR body and `docs/supported-models.md` so a later reader does not undo this. Second, when decode crosses `L`, transformers drops the cache and re-encodes the whole sequence with the long table; the flip-in-place this issue asks for keeps the short-table keys in the cache and is an approximation of that, so transformers parity is not claimed for a generation that crosses `L` mid-decode.

The three real-checkpoint gates (Phi-3.5-mini parity below and above 4096 tokens, and the phi-3-mini regression control) are left unchecked and are run outside the implementation unit.

**Comment by inureyes** [comment]
Follow-up on the correction above: the first version of the fix announced the prompt span in the two chunked-prefill functions only, and that is not where all of the server's prefill forwards live. With the prompt cache on, which is the default, one prompt also reaches the model through `capture_history_boundary_snapshot` (forwards `prompt_tokens[start..boundary]`, a strict prefix that can end below the threshold while the prompt does not) and through `run_next_prompt_cache_warmup` (forwards the delta after restoring a snapshot, where the restored prefix is not in the scheduler's `KVCache::offset` for a non-batching family). Both were un-announced, so `mlxcel-server` returned repetition for a 5136-token prompt through POST /v1/completions at the default 512-token chunk while every CLI gate on the same checkpoint passed.

The announcement is now made once per scheduler entry point that runs prefill work for one sequence, through a `BatchScheduler::announce_prefill_span(&seq)` helper, rather than once per forward. Function scope is the right lifetime: the scheduler interleaves other sequences' decode batches between two chunks of one prompt, but everything inside one of those calls belongs to one sequence. `run_padded_batched_prefill` stays deliberately un-announced (offset 0, span equals the cohort's longest row, so the pass already carries the answer for that row) and the decode forwards must not announce at all.

`src/server/batch/scheduler/prefill_span_coverage_tests.rs` turns that classification into a test that reads the scheduler sources: a forward added to a function the table does not name fails until someone classifies it, and a prefill function that stops announcing fails too. Reaching those forwards from a unit test needs a live `BatchScheduler` and a real model, so nothing else in `cargo test` can see the omission. Two behavioral tests cover the same ground from the model side, one walking the exact geometry that regressed and one driving `chunked_prefill_last_logits` with a probe model that reads `prefill_span::current()` inside `forward`. The full forward-by-forward audit is in the PR body.

**Comment by inureyes** [comment]
Third round, and the gap this time was one level further out than the last two. Announcing the prefill span covers every path that encodes one prompt; it does not touch the path that restores KV encoded by a different one. With the radix prefix cache on, a 1078-token request correctly encoded its prompt under the short table and donated it, and a 5136-token request whose prompt starts with those same tokens hit `cached=1056/5136` and prefilled only the remaining 4080 under the long table. A third identical request then hit `cached=5136/5136` and decoded from the poisoned entry.

A position-selected frequency table makes a cached prefix valid only for requests on the same side of `original_max_position_embeddings`, and it cuts both ways: a prefix a long request encoded must not be restored into a short one either. The fix is `LanguageModel::rope_table_regime(total_prompt_len) -> Option<u8>`, defaulting to `None` for every model whose rotation does not depend on the sequence length. `Phi3Model` answers it by reading layer 0's LongRoPE tables rather than a copy of the config, so the cache's tag cannot drift from what the attention layers rotate with, and a checkpoint with no `longrope` block answers `None` and keeps its reuse untouched. `LoadedModel` delegates it and the VLM wrapper forwards it to the text model, so Phi-3 Vision, Phi4MM and Phi4 SigLIP inherit it.

The tag joins prompt-cache bucket identity in all three places a lookup can reach an entry: the `PromptCacheKey` digest (domain separator bumped to v3 so a v2 entry cannot be read as regime-free), `BucketKey` for the per-session prefix scan, and `SessionlessBucketKey`, the radix trie's index key and the one the observed regression went through. The span the scheduler tags with is the announced prefill span rather than the length of the vector being stored, because a 3000-token history-boundary snapshot cut out of a 5136-token prompt was rotated with that prompt's table. An entry whose generation crossed the boundary mid-decode spans both tables and is declined outright, counted as `rope_regime_mismatch` in `/v1/cache/stats`. A miss falls back to a full prefill under the request's own table, which is what transformers does by resetting its cache; sharing within one regime is unchanged. The cross-node disaggregated handoff needs no tag: it carries no prompt-cache state and moves one request's KV between its own prefill and decode nodes.

The reuse-channel audit and the test list are in the PR body. Worth recording for the issue: all three rounds missed the same invariant one scope out each time, that everything putting keys into one KV cache must agree on the table, whether that is the passes of one prefill, the several places one request reaches the model, or two requests sharing a prefix.
